daml/Marketplace/Clearing/Service.daml

```diff
@@ -0,0 +1,212 @@
+        with
+          targetAmount : Decimal
+          currency : Text
+          calculationId : Text
+        do
+          calculationTime <- getTime
+          create MarginCalculation with ccp = provider; accountId = clearingAccount.id; ..
+
+      nonconsuming CreateMarkToMarket : ContractId MarkToMarketCalculation
+        with
+          mtmAmount : Decimal
+          currency : Text
+          calculationId : Text
+        do
+          calculationTime <- getTime
+          create MarkToMarketCalculation with ccp = provider; accountId = clearingAccount.id; ..
+
+      nonconsuming PerformMarginFill : Optional ((ContractId AssetDeposit), [ContractId AssetDeposit])
+        with
+          depositCids : [ContractId AssetDeposit]
+          marginDepositCids : [ContractId AssetDeposit]
+          calculationCid : ContractId MarginCalculation
+        do
+          targetAmount  <- (.targetAmount) <$> fetch calculationCid
+          depositAmount <- getDepositQuantities depositCids
+          marginAmount  <- getDepositQuantities marginDepositCids
+
+          let diff = targetAmount - marginAmount
+          if diff > 0.0 && depositAmount < diff
+          then resolveMarginCalculation calculationCid CalculationFailed >> return None
+          else do
+            resolveMarginCalculation calculationCid CalculationSatisfied
+            if | diff == 0.0 -> return None
+               | diff > 0.0  -> Some <$> exercise self TransferToMargin with amount = diff; ..
+               | otherwise   -> Some <$> exercise self TransferFromMargin with amount = (abs diff); ..
+
+      nonconsuming PerformMarkToMarket : Optional (ContractId AssetDeposit, [ContractId AssetDeposit])
+        with
+          providerDepositCids : [ContractId AssetDeposit]
+          customerDepositCids : [ContractId AssetDeposit]
+          calculationCid : ContractId MarkToMarketCalculation
+        do
+          mtmAmount      <- (.mtmAmount) <$> fetch calculationCid
+          providerAmount <- getDepositQuantities providerDepositCids
+          customerAmount <- getDepositQuantities customerDepositCids
+
+          let failCalc : Update ()
+              failCalc = resolveMarkToMarketCalculation calculationCid CalculationFailed
+
+              satisfyCalc : Update ()
+              satisfyCalc = resolveMarkToMarketCalculation calculationCid CalculationSatisfied
+
+          if
+            | mtmAmount == 0.0 -> satisfyCalc >> return None
+            | mtmAmount > 0.0  ->
+              if providerAmount < mtmAmount
+              then failCalc >> return None
+              else satisfyCalc >> Some <$> exercise self TransferFromProvider with amount = mtmAmount; depositCids = providerDepositCids
+            | otherwise ->
+              if customerAmount < abs mtmAmount
+              then failCalc >> return None
+              else satisfyCalc >> Some <$> exercise self TransferToProvider with amount = abs mtmAmount; depositCids = customerDepositCids
```

Comment:
Again, this could maybe be tidied up by having a new return type. Consider returning the contractId of the failed calculation in a sum type perhaps? That way you can avoid returning those `Update ()` in the helper functions above, and remove the `>>`.
I guess my main gripe here is that you're side-effecting using `>>`, and it makes the code sort of opaque.
An alternative to changing the return type would be to try and leverage the `Option` monad.

Reply:
Added a return type for this one - split up for the calculation and the remaining deposits, it should be cleaner now. I'll have to experiment with using `Option` as a monad - I've done it a bit in Haskell before with `Maybe` as a monad, makes sense that `Option` will work the same way.